tests/standalone_display_test.cpp

```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright (C) 2022 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// NOLINTNEXTLINE(cppcoreguidelines-macro-usage)
+// #define LOG_NDEBUG 0 // Uncomment to see HWC2 API calls in logcat
+
+#define LOG_TAG "standalone_display_test"
+
+#include <xf86drm.h>
+
+#include <algorithm>
+#include <cinttypes>
+#include "DrmHwcTwo.h"
+#include "HwcLayer.h"
+#include "backend/Backend.h"
+#include "utils/log.h"
+#include <cutils/properties.h>
+#include <string.h>
+#include <sys/mman.h>
+
+namespace android {
+class DrmHwcTwoTest : public DrmHwcTwo {
+ public:
+  DrmHwcTwoTest() {
+    GetResMan().Init();
+  }
+  ~DrmHwcTwoTest() override = default;
+};
+struct Drmhwc2DeviceTest : hwc2_device {
+  DrmHwcTwoTest drmhwctwotest;
+};
+}  // namespace android
+
+typedef struct _ColorInfo {
+  uint8_t A;
+  uint8_t R;
+  uint8_t G;
+  uint8_t B;
+} ColorInfo;
+
+ColorInfo g_colorinfo[2] = {
+    {0x32, 0x80, 0x00, 0 },
+    {0x50, 0x50, 0, 0x20 },
+};
+
+typedef struct buffer_object {
+  uint32_t width;
+  uint32_t height;
+  uint32_t stride;
+  uint32_t handle;
+  uint32_t size;
+  uint8_t *vaddr;
+  uint32_t fb_id;
+
+} Buffer_object;
+/// fb buffer
+struct buffer_object g_bo[2];
+
+int modeset_create_fb(int fd, Buffer_object* bo, uint8_t index);
+void modset_destory_fb(int fd, Buffer_object* bo);
+void fill_buffer(Buffer_object* bo, uint8_t index);
+
+int main(int argc, const char** argv) {
+  ALOGI("usage: %s [buffers num]:2 \n", argv[0]);
+  int buffers_num = 2;
+
+  auto ctx = std::make_unique<android::Drmhwc2DeviceTest>();
+  if (!ctx) {
+    ALOGE("Failed to allocate DrmHwcTwo");
+    return -ENOMEM;
+  }
+
+  android::DrmHwcTwoTest* hwc2 = &ctx->drmhwctwotest;
+  hwc2_display_t display_id = 0;
+  hwc2_layer_t layer_id = 0;
+  int fd = 0;
+
+  while (true) {
```

Comment:
why there is for loop here?
And it is better to put all initialization code to one init function.